src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.EVP.DigestAlgs.cs

```diff
@@ -14,45 +14,107 @@ internal static partial class Crypto
         private static volatile IntPtr s_evpSha256;
         private static volatile IntPtr s_evpSha384;
         private static volatile IntPtr s_evpSha512;
+        private static volatile IntPtr s_evpSha3_256;
+        private static volatile IntPtr s_evpSha3_384;
+        private static volatile IntPtr s_evpSha3_512;
+        private static volatile bool s_evpSha3_256Cached;
+        private static volatile bool s_evpSha3_384Cached;
+        private static volatile bool s_evpSha3_512Cached;
 
         [LibraryImport(Libraries.CryptoNative)]
         private static partial IntPtr CryptoNative_EvpMd5();
 
-        internal static IntPtr EvpMd5() =>
+        private static IntPtr EvpMd5() =>
             s_evpMd5 != IntPtr.Zero ? s_evpMd5 : (s_evpMd5 = CryptoNative_EvpMd5());
 
         [LibraryImport(Libraries.CryptoNative)]
-        internal static partial IntPtr CryptoNative_EvpSha1();
+        private static partial IntPtr CryptoNative_EvpSha1();
 
-        internal static IntPtr EvpSha1() =>
+        private static IntPtr EvpSha1() =>
             s_evpSha1 != IntPtr.Zero ? s_evpSha1 : (s_evpSha1 = CryptoNative_EvpSha1());
 
         [LibraryImport(Libraries.CryptoNative)]
-        internal static partial IntPtr CryptoNative_EvpSha256();
+        private static partial IntPtr CryptoNative_EvpSha256();
 
-        internal static IntPtr EvpSha256() =>
+        private static IntPtr EvpSha256() =>
             s_evpSha256 != IntPtr.Zero ? s_evpSha256 : (s_evpSha256 = CryptoNative_EvpSha256());
 
         [LibraryImport(Libraries.CryptoNative)]
-        internal static partial IntPtr CryptoNative_EvpSha384();
+        private static partial IntPtr CryptoNative_EvpSha384();
 
-        internal static IntPtr EvpSha384() =>
+        private static IntPtr EvpSha384() =>
             s_evpSha384 != IntPtr.Zero ? s_evpSha384 : (s_evpSha384 = CryptoNative_EvpSha384());
 
         [LibraryImport(Libraries.CryptoNative)]
-        internal static partial IntPtr CryptoNative_EvpSha512();
+        private static partial IntPtr CryptoNative_EvpSha512();
 
-        internal static IntPtr EvpSha512() =>
+        private static IntPtr EvpSha512() =>
             s_evpSha512 != IntPtr.Zero ? s_evpSha512 : (s_evpSha512 = CryptoNative_EvpSha512());
 
-        internal static IntPtr HashAlgorithmToEvp(string hashAlgorithmId) => hashAlgorithmId switch
+        [LibraryImport(Libraries.CryptoNative)]
+        private static partial IntPtr CryptoNative_EvpSha3_256();
+
+        private static IntPtr EvpSha3_256()
+        {
+            if (!s_evpSha3_256Cached)
+            {
+                s_evpSha3_256 = CryptoNative_EvpSha3_256();
+                s_evpSha3_256Cached = true;
```

Comment:
Why do we need a separate cached bool field?  Is IntPtr.Zero a valid value?

Reply:
> Is IntPtr.Zero a valid value?

It is a valid value to mean "This openssl version does not support SHA3". We don't have that for SHA2 because all versions of OpenSSL we support also support SHA2.